pkg/cloud/aws/services/ec2/instances.go

```diff
@@ -0,0 +1,305 @@
+			instances = append(instances, instance)
+		}
+	}
+
+	return instances, nil
+}
+
+// GetRunningInstances returns all running instances that have a tag matching our machine name,
+// and cluster ID.
+func (ir *instancesRuntime) GetRunningInstances(machine *clusterv1.Machine) ([]*ec2.Instance, error) {
+	runningInstanceStateFilter := []*string{aws.String(ec2.InstanceStateNameRunning), aws.String(ec2.InstanceStateNamePending)}
+	return ir.getInstances(machine, runningInstanceStateFilter)
+}
+
+// getStoppedInstances returns all stopped instances that have a tag matching our machine name,
+// and cluster ID.
+func (ir *instancesRuntime) getStoppedInstances(machine *clusterv1.Machine) ([]*ec2.Instance, error) {
+	stoppedInstanceStateFilter := []*string{aws.String(ec2.InstanceStateNameStopped), aws.String(ec2.InstanceStateNameStopping)}
+	return ir.getInstances(machine, stoppedInstanceStateFilter)
+}
+
+// TerminateInstances terminates all provided instances with a single EC2 request.
+func (ir *instancesRuntime) TerminateInstances(instances []*ec2.Instance) error {
+	instanceIDs := []*string{}
+	// Cleanup all older instances:
+	for _, instance := range instances {
+		glog.Infof("Cleaning up extraneous instance for machine: %v, state: %v, launchTime: %v", *instance.InstanceId, *instance.State.Name, *instance.LaunchTime)
+		instanceIDs = append(instanceIDs, instance.InstanceId)
+	}
+	for _, instanceID := range instanceIDs {
+		glog.Infof("Terminating %v instance", *instanceID)
+	}
+
+	terminateInstancesRequest := &ec2.TerminateInstancesInput{
+		InstanceIds: instanceIDs,
+	}
+	_, err := ir.awsclient.TerminateInstances(terminateInstancesRequest)
+	if err != nil {
+		glog.Errorf("error terminating instances: %v", err)
+		return fmt.Errorf("error terminating instances: %v", err)
+	}
+	return nil
+}
+
+// RemoveStoppedMachine removes all instances of a specific machine that are in a stopped state.
+func (ir *instancesRuntime) RemoveStoppedMachine(machine *clusterv1.Machine) error {
+	instances, err := ir.getStoppedInstances(machine)
+	if err != nil {
+		return fmt.Errorf("unable to remove stopped instances, error getting stopped instances: %v", err)
+	}
+
+	if len(instances) == 0 {
+		glog.Infof("no stopped instances found for machine %v", machine.Name)
+		return nil
+	}
+
+	return ir.TerminateInstances(instances)
+}
+
+// RemoveRunningMachine removes all instances of a specific machine that are in a running state.
+func (ir *instancesRuntime) RemoveRunningMachine(machine *clusterv1.Machine) error {
+	instances, err := ir.GetRunningInstances(machine)
+	if err != nil {
+		return fmt.Errorf("unable to remove running instances, error getting running instances: %v", err)
+	}
+
+	if len(instances) == 0 {
+		glog.Infof("no running instances found for machine %v", machine.Name)
+		return nil
+	}
+
+	return ir.TerminateInstances(instances)
+}
+
+// LaunchConfig configuration for launching instances
+type LaunchConfig struct {
```

Comment:
What does this gives us? for an aws specific there's already `RunInstancesInput` Why not just use providerconfigv1? I can't see how we can benefit from this abstraction. If any I would abstract the logic for getting the final input based on the user input, e. ami id (181..215) or how we build the tagList so then those can be unit tested.

Reply:
What about:
```go
type LaunchConfig struct {
	ProviderConfig *providerconfigv1.AWSMachineProviderConfig
	UserData []byte
}
```
then? The things is the `services/ec2` packages is meant to abstract from ec2 as much as possible and offer API based on machines so one can just provide a machine object without any expectations on its content. Plus, we need a way to wire user data which are collected from a k8s secret. Just trying to make the methods kube client unaware.

Reply:
This idea is dropped as the `LaunchConfig` is removed.